sci-biology/imagej/imagej-1.54d.ebuild

```diff
@@ -0,0 +1,128 @@
+# Copyright 1999-2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit java-pkg-2 java-ant-2 desktop
+
+MY_PN="ij"
+IJ_PV="153" #plugins not currently available under 154
+
+DESCRIPTION="Image Processing and Analysis in Java"
+
+HOMEPAGE="
+	https://imagej.nih.gov/ij/
+	https://github.com/imagej
+"
+
+SRC_URI="
+	https://github.com/imagej/ImageJ/archive/refs/tags/v${PV}.tar.gz -> ${P}.gh.tar.gz
+	https://rsb.info.nih.gov/ij/images/ImageJ.png
+	plugins? ( https://wsr.imagej.net/distros/cross-platform/${MY_PN}${IJ_PV}.zip )"
+# plugins are under a different licenses and can be installed into user's $IJ_HOME/plugins
+#	plugins? ( http://rsb.info.nih.gov/ij/download/zips/${MY_PN}${IJ_PV}.zip )"
+
+KEYWORDS="~amd64"
+
+S="${WORKDIR}/ImageJ-${PV}"
+
+LICENSE="public-domain"
+
+SLOT="0"
+
+IUSE="doc plugins debug"
+
+RDEPEND="
+	>=virtual/jre-1.7:*
+	dev-java/java-config
+"
+DEPEND="${RDEPEND}
+	>=virtual/jdk-1.7:*
+"
+BDEPEND="
+	dev-java/ant-core
+	app-arch/unzip
+"
+
+IJ_S=${WORKDIR}/ImageJ
+
+src_prepare() {
+	cp "${DISTDIR}"/ImageJ.png "${WORKDIR}/${PN}.png" || die
+
+	if ! use debug ; then
```

Comment:
Don't we need the `if use plugins` stuff from the other ebuild here as well?

Reply:
Hi @AndrewAmmerlaan 
I am not sure it is needed, but in the imagej-1.54d ebuild, the SRC_URI is actually both fetching and unpacking the ImageJ zip sources and unpacking them correctly. 

It only appeared necessary for the 9999 ebuild, which after inheriting the git-r3 class was no longer unpacking every source listed on the SRC_URI

In other others words; No. It did not seem necessary, except for the 9999 to explicitly unpack.

Reply:
> It only appeared necessary for the 9999 ebuild, which after inheriting the git-r3 class was no longer unpacking every source listed on the SRC_URI

This happens because the `git-r3.eclass` comes with its own `git-r3_src_unpack()` overriding the default `src_unpack()`. If you want to use both, you'll need to do something like this:
```
src_unpack() {
    default
    git-r3_src_unpack
}
```

I would suggest structuring both ebuilds like this:
```
inherit ....

DESCRIPTION=...
HOMEPAGE=...
SRC_URI=.... # The png and the plugins

if [[ ${PV} == 9999 ]]; then
    inherit git-r3
    EGIT_REPO_URI=....
    S=.....
else
    SRC_URI+=.... # The main sources
    S=.... 
    KEYWORDS="~amd64"
fi

src_unpack() {
    default
    if [[ ${PV} == 9999 ]]; then
         git-r3_src_unpack
    fi

.....

```
This way both ebuild files can be identical, this simplifies maintenance later on, for the next release you can just update the 9999, copy and rename it.

Reply:
Interesting ebuild structure. I surely don't hate it, but also I surely don't thing it necessarily simplifies maintenance. 

As the ebuilds currently look, the unpack is only being called for the situation that appear to require it. That is, when the default unpack worked graciously (in 1.53 and 1.54), then nothing was added. 
Say, KISS.

On the maintenance issue, it seems the current situation is a pretty straightfoward case:

1. for 1.53: Rename ebuild accordingly when the NIH releases 1.54 zip sources
2. for new tag releases in github: Rename ebuild accordingly when releases are Tagged in the repository
3. for the plugins in the 1.54 and 9999 versions: Update the variables that control the version of plugins on these ebuilds. 

But I would admit, here, we enter a bit on stylistic issues; but also a lot more on an experience situation. Therefore, my initial reaction being: I don't concur. But I can be persuaded to the contrary.

Reply:
> On the maintenance issue, it seems the current situation is a pretty straightfoward case:
> 
>     1. for 1.53: Rename ebuild accordingly when the NIH releases 1.54 zip sources
> 
>     2. for new tag releases in github: Rename ebuild accordingly when releases are Tagged in the repository
> 
>     3. for the plugins in the 1.54 and 9999 versions: Update the variables that control the version of plugins on these ebuilds.

Yes this is one way to do it, we go from old, copy to new and then make changes. The argument to do it the other way around is that changes always land in 9999 first. Therefore if 9999 had to be changed in between releases you don't have to make this same change again in the next release if you copy from 9999 instead of from the previous release. For this to work 9999 has to be written in a way that the git stuff is behind if statements.

The second argument to use the if statements is that the base structure of the ebuilds is then basically the same, this way you can easily spot changes in configure/install phases between different releases and live.

This is the way we usually do it, if you take a look at ::gentoo you'll see that most/all of the live ebuilds use this if statement structure. That being said, if you're maintaining it you decide so feel free to use a different workflow.

Reply:
Ok. given the method you outline is the Gentoo way for most/all live ebuilds, I shall go ahead and try if I can make the ebuilds work with the structure proposed. 
I'll let you know if I can (or not) make it work.

Regards

Reply:
https://github.com/gentoo/sci/pull/1198/commits/365a963f3d5b436e0961cdd70b1fd7e3bc5770e9 

Restructures the new ebuilds as suggested, versions 1.54d and 9999, which fetch from github, are the same file copied over. Conditional statements on the ebuild handle the differences between versions.

Regards